### ISSUE TITLE
Unused Audio Tokens?

### ISSUE THREAD

**Issue by Myzomi** [canopyai/Orpheus-TTS#200]
From how I understand this, Orpheus appends 28682 audio tokens labeled <custom_token_{id}> with:
`
number_add_tokens = 7 * 4096 + 10
`
`
new_tokens = [f"<custom_token_{i}>" for i in range(0, number_add_tokens + 1)]
`

where 7 is the number of codebooks , 4096 is the size of each codebook, and 10 for sentinel tokens (8 used, 2 unused)
However in the code, it is appending 28682 + 1 tokens. `[0, ..., 28681,28682] `

Along with this, the vocab size inside of canopylabs/orpheus-3b is 156940

if you subtract 156940 - 128256 (meta-llama/Llama-3.2-3B's vocab size)
you get, 128258 which is off by 2 tokens.

one can be explained by the single unused token but what is the other token? another pad token?

**Comment by amuvarma13** [comment]
We are leaving a couple of tokens as spare tokens if we need them for other things, i.e. if we want to finetune the model to do global emotion we could have these as start_of_global_emotion end_of_global etc.

**Comment by saifulislam79** [comment]
@amuvarma13 thanks for your explanation, but SNAC cookbook is **4096** and it's Downsampling, Upsampling  each quantization stage, the residual is downsampled by a factor (e.g., 2, 4, 8), quantized using a codebook. so all token cover into 4096. For your speech token if we add **meta-llama/Llama-3.2-3B's vocab size**)+ **Number of codebook(4096)**+ 10 for sentinel tokens(as you said).
why need to add number_add_tokens = **7 * 4096** and the SNAC audio code defined each frame constrain 7 token but this token not more then 4096 of it cookbook.

**Comment by saifulislam79** [comment]
@amuvarma13 and other question is why need remove **remove_duplicate_frames** of it data processing code.
#############################################
```
# 3. Data loading and processing functions
#############################################
def remove_duplicate_frames(example):
    vals = example["codes_list"]
    if len(vals) % 7 != 0:
        raise ValueError("Input list length must be divisible by 7")

    result = vals[:7]

    removed_frames = 0

    for i in range(7, len(vals), 7):
        current_first = vals[i]
        previous_first = result[-7]

        if current_first != previous_first:
            result.extend(vals[i:i+7])
        else:
            removed_frames += 1

    example["codes_list"] = result
    return example
```

i have analysis SNAC audio code convert the audio quantization structure to list of token and reconstruct it quantization and got audio as input but why need remove_duplicate_frames for codec_list process while i want to train orpheus-tts.

**Comment by amuvarma13** [comment]
Leads to stabler training.

**Comment by amuvarma13** [comment]
> [@amuvarma13](https://github.com/amuvarma13) thanks for your explanation, but SNAC cookbook is **4096** and it's Downsampling, Upsampling each quantization stage, the residual is downsampled by a factor (e.g., 2, 4, 8), quantized using a codebook. so all token cover into 4096. For your speech token if we add **meta-llama/Llama-3.2-3B's vocab size**)+ **Number of codebook(4096)**+ 10 for sentinel tokens(as you said). why need to add number_add_tokens = **7 * 4096** and the SNAC audio code defined each frame constrain 7 token but this token not more then 4096 of it cookbook.

Because token label "1" or the first layer isn't the same as token label "1" of the 3rd layer so we don't want the LLM to confuse these, when it creates input_embeddings

**Comment by saifulislam79** [comment]
@amuvarma13 basically here the processing code found,
```
           all_codes.append(codes[0][0][i].item()+128266)
            all_codes.append(codes[1][0][2*i].item()+128266+4096)
            all_codes.append(codes[2][0][4*i].item()+128266+(2*4096))
            all_codes.append(codes[2][0][(4*i)+1].item()+128266+(3*4096))
            **all_codes.append(codes[1][0][(2*i)+1].item()+128266+(4*4096))**
            all_codes.append(codes[2][0][(4*i)+2].item()+128266+(5*4096))
            all_codes.append(codes[2][0][(4*i)+3].item()+128266+(6*4096))
```

why you assign 4 index of 1 index(**all_codes.append(codes[1][0][(2*i)+1].item()+128266+(4*4096))**) codec and not maintenance serial or incremental list like below,
```
    levels = [c.squeeze(0) for c in codes]  # each is [T_i]

    T = levels[0].shape[0]  # base resolution
    assert levels[1].shape[0] == 2 * T
    assert levels[2].shape[0] == 4 * T

    tokens = []
    for t in range(T):
        tokens.append(levels[0][t])
        tokens.append(levels[1][2 * t])
        tokens.append(levels[1][2 * t + 1])
        tokens.append(levels[2][4 * t])
        tokens.append(levels[2][4 * t + 1])
        tokens.append(levels[2][4 * t + 2])
        tokens.append(levels[2][4 * t + 3])
    return torch.stack(tokens)
```
i will great full to you, please give your answer.
